Kernel/Graphics/VirtIOGPU/VirtIOGPU.cpp

```diff
@@ -0,0 +1,286 @@
+        draw_test_pattern();
+        // 6. Use VIRTIO_GPU_CMD_TRANSFER_TO_HOST_2D to update the host resource from guest memory.
+        transfer_framebuffer_data_to_host(m_display_info.rect);
+        // 7. Use VIRTIO_GPU_CMD_RESOURCE_FLUSH to flush the updated resource to the display.
+        flush_displayed_image(m_display_info.rect);
+    } else {
+        VERIFY_NOT_REACHED();
+    }
+}
+
+VirtIOGPU::~VirtIOGPU()
+{
+}
+
+bool VirtIOGPU::handle_device_config_change()
+{
+    return false;
+}
+
+void VirtIOGPU::handle_queue_update(u16 queue_index)
+{
+    dbgln_if(VIRTIO_DEBUG, "VirtIOGPU: Handle queue update");
+    VERIFY(queue_index == CONTROLQ);
+
+    auto& queue = get_queue(CONTROLQ);
+    ScopedSpinLock queue_lock(queue.lock());
+    queue.discard_used_buffers();
+    m_has_outstanding_request = false;
+}
+
+u32 VirtIOGPU::get_pending_events()
+{
+    return config_read32(*m_device_configuration, DEVICE_EVENTS_READ);
+}
+
+void VirtIOGPU::clear_pending_events(u32 event_bitmask)
+{
+    config_write32(*m_device_configuration, DEVICE_EVENTS_READ, event_bitmask);
+}
+
+void VirtIOGPU::query_display_information()
+{
+    Locker locker(m_operation_lock);
+    auto& request = *reinterpret_cast<VirtIOGPUCtrlHeader*>(m_scratch_space->vaddr().as_ptr());
+    auto& response = *reinterpret_cast<VirtIOGPURespDisplayInfo*>((m_scratch_space->vaddr().offset(sizeof(request)).as_ptr()));
+
+    populate_virtio_gpu_request_header(request, VirtIOGPUCtrlType::VIRTIO_GPU_CMD_GET_DISPLAY_INFO, VIRTIO_GPU_FLAG_FENCE);
+
+    synchronous_virtio_gpu_command(start_of_scratch_space(), sizeof(request), sizeof(response));
+
+    for (size_t i = 0; i < VIRTIO_GPU_MAX_SCANOUTS; ++i) {
+        auto& scanout = response.scanout_modes[i];
+        if (!scanout.enabled)
+            continue;
+        dbgln("Scanout {}: x: {}, y: {}, width: {}, height: {}", i, scanout.rect.x, scanout.rect.y, scanout.rect.width, scanout.rect.height);
+        m_display_info = scanout;
+        m_chosen_scanout = i;
+    }
+    m_display_info.rect.width = 1024;
+    m_display_info.rect.height = 768;
+}
+
+void VirtIOGPU::create_framebuffer()
+{
+    Locker locker(m_operation_lock);
+    auto& request = *reinterpret_cast<VirtIOGPUResourceCreate2D*>(m_scratch_space->vaddr().as_ptr());
+    auto& response = *reinterpret_cast<VirtIOGPUCtrlHeader*>((m_scratch_space->vaddr().offset(sizeof(request)).as_ptr()));
+
+    populate_virtio_gpu_request_header(request.header, VirtIOGPUCtrlType::VIRTIO_GPU_CMD_RESOURCE_CREATE_2D, VIRTIO_GPU_FLAG_FENCE);
+
+    request.resource_id = m_framebuffer_id = 1;
+    request.width = m_display_info.rect.width;
+    request.height = m_display_info.rect.height;
+    request.format = static_cast<u32>(VirtIOGPUFormats::VIRTIO_GPU_FORMAT_B8G8R8X8_UNORM);
+
+    synchronous_virtio_gpu_command(start_of_scratch_space(), sizeof(request), sizeof(response));
+
+    VERIFY(response.type == static_cast<u32>(VirtIOGPUCtrlType::VIRTIO_GPU_RESP_OK_NODATA));
```

Comment:
should we really VERIFY here? device failure usually shouldnt result in whole system failure.

Reply:
That might be true for stuff like flushing the buffer, but if we have issues attaching the scanouts, I'm not sure what graceful recovery would be in this scenario. I'm not 100% sure which way to go for this one.

Reply:
I think we should just panic at this point. If the commands start failing, it will be really hard to recover from, and without a way to ssh into the Serenity VM I guess we'd be SOL. :/ I'm going to leave these in, because I really can't think of a good way to recover from this.